### PR TITLE
ingest: correctly measure chained pipeline stats

### PR DESCRIPTION
**Pull request by jakelandis**
Prior to this change when a pipeline processor called another
pipeline, only the stats for the first processor were recorded.
The stats for the subsequent pipelines were ignored. This change
properly accounts for pipelines irregardless if they are the first
or subsequently called pipelines.

This change moves the state of the stats from the IngestService
to the pipeline itself. Cluster updates are safe since the pipelines
map is atomically swapped, and if a cluster update happens
while iterating over stats (now read directly from the pipeline)
a slightly stale view of stats may be shown.

-------------

With the following setup:

```
PUT _ingest/pipeline/mypipeline1
{
  "processors": [
    {
      "pipeline": {
        "pipeline": "mypipeline2"
      }
    }
  ]
}

PUT _ingest/pipeline/mypipeline2
{
  "processors": [
    {
      "pipeline": {
        "pipeline": "mypipeline3"
      }
    }
  ]
}

PUT _ingest/pipeline/mypipeline3
{
  "processors": [
    {
      "rename": {
        "field": "a",
        "target_field": "a2"
      }
    }
  ]
}


POST test/_doc?pipeline=mypipeline1
{
  "a" : "b"
}

DELETE _ingest/pipeline/xpack_monitoring_2
DELETE _ingest/pipeline/xpack_monitoring_6
GET _nodes/stats/ingest?filter_path=nodes.*.ingest
```
Before this change:
```
{
  "nodes": {
    "HvZVZ2o0SYqv4pMM8mX1cA": {
      "ingest": {
        "total": {
          "count": 1,
          "time_in_millis": 4,
          "current": 0,
          "failed": 0
        },
        "pipelines": {
          "mypipeline3": {
            "count": 0,
            "time_in_millis": 0,
            "current": 0,
            "failed": 0
          },
          "mypipeline2": {
            "count": 0,
            "time_in_millis": 0,
            "current": 0,
            "failed": 0
          },
          "mypipeline1": {
            "count": 1,
            "time_in_millis": 4,
            "current": 0,
            "failed": 0
          }
        }
      }
    }
  }
}
```
^^ Notice that pipeline2/3 were called but the stats do not reflect that they are. 
After this change:
```
{
  "nodes": {
    "qdfWAkVhRQSKu8Au3s32fA": {
      "ingest": {
        "total": {
          "count": 1,
          "time_in_millis": 3,
          "current": 0,
          "failed": 0
        },
        "pipelines": {
          "mypipeline1": {
            "count": 1,
            "time_in_millis": 1,
            "current": 0,
            "failed": 0
          },
          "mypipeline3": {
            "count": 1,
            "time_in_millis": 1,
            "current": 0,
            "failed": 0
          },
          "mypipeline2": {
            "count": 1,
            "time_in_millis": 1,
            "current": 0,
            "failed": 0
          }
        }
      }
    }
  }
}
```

Note, before and after this change the time_in_millis includes the children pipelines. It is difficult to emulate outside of unit testing ... but if pipeline3 takes 3ms, and pipeline2 takes 2 milliseconds, and pipeline1 takes 1 ms, (assuming pipeline1->pipelin2->pipeline3) then pipeline1 = 6ms (1 + 2 + 3) , pipeline 2 = 5ms (2 + 3), and pipeline 3ms. 

----------

Also apologizes for the noise around the imports, but if understand correctly the updated versions is the preferred order. (and it is kinda painful to manage them manually or continually change my IDE settings)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 1]
```diff
@@ -367,37 +375,11 @@ protected void doRun() {
     }
 
     public IngestStats stats() {
-        Map<String, StatsHolder> statsHolderPerPipeline = this.statsHolderPerPipeline;
 
-        Map<String, IngestStats.Stats> statsPerPipeline = new HashMap<>(statsHolderPerPipeline.size());
-        for (Map.Entry<String, StatsHolder> entry : statsHolderPerPipeline.entrySet()) {
-            statsPerPipeline.put(entry.getKey(), entry.getValue().createStats());
-        }
+        Map<String, IngestStats.Stats> statsPerPipeline =
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java]: NIT: I think even in `2018` JDKs the old version with the manual iteration is faster. I don't think it matters much here, but probably not worth the noise to change to the functional iteration here.

File: server/src/test/java/org/elasticsearch/ingest/PipelineProcessorTests.java [hunk 2]
```diff
@@ -130,4 +126,76 @@ public void testAllowsRepeatedPipelineInvocations() throws Exception {
...
+        );
+
+        String key1 = randomAlphaOfLength(10);
+        Pipeline pipeline2 = new Pipeline(
+            pipeline2Id, null, null, new CompoundProcessor(true,
+            Arrays.asList(
+                new TestProcessor(ingestDocument -> {
+                    ingestDocument.setFieldValue(key1, randomInt());
+                    try {
+                        Thread.sleep(100); //force the stat time to be non-zero
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/ingest/PipelineProcessorTests.java]: Shouldn't it technically suffice to sleep for `2` here (or even `1`, not sure about the potential rounding/accuracy issues with that and `nanotime()`, but `2` should be fine for getting a visible change in the metric?)

**Review by original-brownbear — COMMENTED** [review]
just a few nits :)
If @rjernst is happy I'm happy

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 3]
```diff
@@ -77,10 +76,9 @@
     // We know of all the processor factories when a node with all its plugin have been initialized. Also some
     // processor factories rely on other node services. Custom metadata is statically registered when classes
     // are loaded, so in the cluster state we just save the pipeline config and here we keep the actual pipelines around.
-    private volatile Map<String, Pipeline> pipelines = new HashMap<>();
+    private volatile Map<String, Pipeline> pipelines = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: Ah.. you are right updates are rebuilt local to a method then atomically swapped to an unmodified map.  => no need for CHM 

will put back to normal HM and update commit message.

File: server/src/test/java/org/elasticsearch/ingest/PipelineProcessorTests.java (hunk 2 above)
Comment (jakelandis) [reply server/src/test/java/org/elasticsearch/ingest/PipelineProcessorTests.java]: yeah, 100 is bit overkill. I will drop this down.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 4]
```diff
@@ -257,10 +255,19 @@ public IngestInfo info() {
     @Override
     public void applyClusterState(final ClusterChangedEvent event) {
         ClusterState state = event.state();
+        int beforeHashCode = pipelines.hashCode();
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: agreed. now that i see the update is is an atomic swap I can clean this up by holding onto the original reference.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 1 above)
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: variable name is the same, but the object and how it iterates is different, hence the update. (e.g. more the noise)

**Comment by jakelandis** [comment]
@original-brownbear thanks for the review. changes added on https://github.com/elastic/elasticsearch/pull/33912/commits/526ae4b381fe764b0f91677871c394011e2535fc

**Comment by jakelandis** [comment]
Jenkins test this please.

**Comment by jakelandis** [comment]
@rjernst - ping

**Comment by jakelandis** [comment]
Jenkins test this please.

**Review by rjernst — COMMENTED** [review]
I have a couple questions

File: server/src/main/java/org/elasticsearch/ingest/IngestMetric.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+package org.elasticsearch.ingest;
+
+import org.elasticsearch.common.metrics.CounterMetric;
+import org.elasticsearch.common.metrics.MeanMetric;
+
+class IngestMetric {
+
+    private final MeanMetric ingestMetric = new MeanMetric();
+    private final CounterMetric ingestCurrent = new CounterMetric();
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestMetric.java]: `current` is briefly incremented while the ingest thing that is getting measured (total, pipeline, (soon) processor) is working on a document. It makes the most sense for total to see how many documents are currently being processed. 

I will add some Javadoc.

**Comment by jakelandis** [comment]
@rjernst - changes implemented, can you take another look ?

File: server/src/main/java/org/elasticsearch/ingest/IngestMetric.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        ingestMetric.inc(ingestTimeInMillis);
+        ingestCount.inc();
+    }
+
+    void ingestFailed() {
+        ingestFailed.inc();
+    }
+
+    void add(IngestMetric metrics){
+        ingestCount.inc(metrics.ingestCount.count());
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply server/src/main/java/org/elasticsearch/ingest/IngestMetric.java]: hopefully the added javadoc sufficiently addresses the missing addition of `current`.

File: server/src/main/java/org/elasticsearch/ingest/IngestMetric.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+ * This class does not make assumptions about it's given scope.
+ */
+class IngestMetric {
+
+    /**
+     * The time it takes to complete the measured item.
+     */
+    private final MeanMetric ingestTime = new MeanMetric();
+    /**
+     * The current count of things being measure. Should mostly like ever be 0 or 1.
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/ingest/IngestMetric.java]: wording is confusing here: most likely?

File: server/src/main/java/org/elasticsearch/ingest/IngestMetric.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Call this if the ingest action failed.
+     */
+    void ingestFailed() {
+        ingestFailed.inc();
+    }
+
+    /**
+     * Add two sets of metrics together. *Important* does NOT add the current count intentionally,
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/ingest/IngestMetric.java]: I would move this important section down to its own paragraph, and please use proper javadoc formatting for emphasis and paragraph breaks.

**Review by rjernst — APPROVED** [review]
Thanks, LGTM

**Comment by jakelandis** [comment]
6.5 backport: https://github.com/elastic/elasticsearch/commit/8e72a68e52b2225dffd3ab64b73fbb774144b7c2

**Pull request merged**